### ISSUE TITLE
Don't use EFI_RUNTIME_SERVICES->Reboot

### ISSUE THREAD

**Issue by iamahuman** [QubesOS/qubes-issues#6258]
**The problem you're addressing (if any)**
Some buggy firmwares do not correctly implement the `Reboot` method of `EFI_RUNTIME_SERVICES` interface, when the EFI boot services are unmapped.
The end result _usually_ is reboot anyway, since an incorrect implementation would result in a x86 triple fault which, in turn, leads to a (hard) reboot.
I do not however believe this is the most reliable and optimal behaviour.

**Describe the solution you'd like**
- Add `reboot=acpi` to Xen command line.

**Where is the value to a user, and who might that user be?**
- More reliable system reboot
- Soft reboot works on such faulty firmwares, which means the user do not have to re-type boot passwords if the firmware is configured as such.

**Describe alternatives you've considered**
- Documentation may be appropriate, but I'm afraid some users may not even notice this bug.

**Additional context**
https://github.com/torvalds/linux/blob/master/arch/x86/platform/efi/quirks.c#L566-L581

The LKML threads in the following section claims that Windows seems to always use the ACPI reboot method regardless of EFI presence.

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**
- https://lkml.org/lkml/2019/4/17/788
- https://lkml.org/lkml/2019/9/13/864

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**
- #6092

**Comment by iamahuman** [comment]
Ideally this is best fixed upstream)

**Comment by iamahuman** [comment]
@pwmarcz You said in https://github.com/QubesOS/qubes-issues/issues/6066#issuecomment-701298925 that rebooting did not work in one of your laptops. Can you confirm if rebooting still freezes on it with `reboot=acpi`?

**Comment by pwmarcz** [comment]
@iamahuman Yes, after `reboot=acpi` the computer reboots correctly.

**Comment by noskb** [comment]
It also works for me using the ASUS H370-A motherboard(CSM disabed)

Before adding `reboot=acpi`, the reboot was working, but HDD power was forcibly turned off and the  S.M.A.R.T. attribute Power-off Retract Count was increased. It would indicate that the system did not successfully reboot I guess.
>The end result usually is reboot anyway, since an incorrect implementation would result in a x86 triple fault which, in turn, leads to a (hard) reboot.

Maybe like this.

**Comment by iamahuman** [comment]
@andrewdavidwong Should this issue have the `hardware support` tag?

Theoretically, there's nothing wrong with using the EFI interface; rather, this is a workaround for yet another instance of a problem arising from faulty firmwares, which Qubes has encountered countless number of times from the beginning.

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong Should this issue have the `hardware support` tag?
> 
> Theoretically, there's nothing wrong with using the EFI interface; rather, this is a workaround for yet another instance of a problem arising from faulty firmwares, which Qubes has encountered countless number of times from the beginning.

Thanks, added!

**Comment by DemiMarie** [comment]
Does ACPI have any advantages over deliberately triple faulting the processor?

**Comment by iamahuman** [comment]
@DemiMarie In fact, it's not even clear that a triple fault is being done behind the scenes. It may also have other harmful effects resulting from undefined behavior.

In case there is a _strong_ reason to prefer triple faults _despite all this_, then it might be a good idea to simply specify it explicitly instead.

**Comment by iamahuman** [comment]
ACPI allows the firmware to do graceful shutdown and/or warm reboot. Since Windows/Linux does the same on a bare-metal hardware, we can safely assume other users would be expecting this behavior.

The solution is to synchronize the Linux x86 reboot quirks code to Xen. I'm not quite sure about other possible breakages resulting from this, though.

**Comment by miczyg1** [comment]
@marmarek it looks like the issue is also present on Novacustom laptops. However, Dasharo firmware uses the same reset method as declared in ACPI FADT: https://github.com/Dasharo/edk2/blob/dasharo/UefiPayloadPkg/Library/ResetSystemLib/ResetSystemLib.c#L77

coreboot always reports the reset register in FADT as 0xcf9 with standard reset values. UEFI Payload picks it up for the RS reset handlers as in the link above.

tl;dr ACPI reboot and UEFI RS reboot should be equivalent.

The only possible issue may be somewhere between the RS->Reboot call from Xen and the IO Write 0xcf9 in UEFI Paylaod. Is it possible to get a trace from Xen in a case where it hangs so I could pinpoint the location of the fault?

BTW it doesn't seem to happen on MSI boards.

**Comment by marmarek** [comment]
> Is it possible to get a trace from Xen in a case where it hangs so I could pinpoint the location of the fault?

it will be tricky, as I don't have real console there, but I can try.

**Comment by miczyg1** [comment]
> > Is it possible to get a trace from Xen in a case where it hangs so I could pinpoint the location of the fault?
> 
> it will be tricky, as I don't have real console there, but I can try.

Thank you. Just asking for a possibility so we could also debug it on our side. As a last resort, we could solder wires to the UART from PCH to get a trace (MMIO UART should work unless it is deinitialized completely before reboot)

**Comment by marmarek** [comment]
@miczyg1 I haven't got a chance to collect logs from NovaCustom laptop yet, but I got similar crash on MSI: https://gitlab.com/xen-project/people/marmarek/xen/-/jobs/6144779899:
<details><summary>Details</summary>
<p>

```
(XEN) [    5.040906] Hardware Dom0 crashed: rebooting machine in 5 seconds.
(XEN) [   12.449137] ----[ Xen-4.19-unstable  x86_64  debug=y  Not tainted ]----
(XEN) [   12.456945] CPU:    0
(XEN) [   12.459989] RIP:    e008:[<0000000045f1e11f>] 0000000045f1e11f
(XEN) [   12.466942] RFLAGS: 0000000000010206   CONTEXT: hypervisor
(XEN) [   12.473513] rax: 0000000000000033   rbx: 0000000000000000   rcx: 00000000465fe000
(XEN) [   12.482278] rdx: 0000000000000033   rsi: ffff8304a4817a90   rdi: 0000000000000001
(XEN) [   12.491037] rbp: ffff8304a4817950   rsp: ffff8304a4817910   r8:  0000000000000001
(XEN) [   12.499799] r9:  ffff8304a4817990   r10: 0000000000000002   r11: 0000000000000000
(XEN) [   12.508560] r12: 0000000000000033   r13: 0000000000000000   r14: 0000000000000000
(XEN) [   12.517320] r15: 00000000000000fb   cr0: 0000000080050033   cr4: 0000000000f526e0
(XEN) [   12.526081] cr3: 00000004a23ff000   cr2: 00000000465fe004
(XEN) [   12.532557] fsb: 0000000000000000   gsb: ffff8881b9dc0000   gss: 0000000000000000
(XEN) [   12.541322] ds: 002b   es: 002b   fs: 0000   gs: 0000   ss: e010   cs: e008
(XEN) [   12.549514] Xen code around <0000000045f1e11f> (0000000045f1e11f):
(XEN) [   12.556848]  2d 00 00 48 85 c9 74 5e <8b> 71 04 8b 01 89 f3 81 e3 ff ff ff 0f 39 c3 73
(XEN) [   12.566083] Xen stack trace from rsp=ffff8304a4817910:
(XEN) [   12.572273]    0000000000000000 80000000a4817fff 0000000000000000 0000000000000000
(XEN) [   12.581130]    0000000000000000 0000000000000000 ffff8304a4817a90 ffff8304a4817990
(XEN) [   12.589986]    ffff8304a4817b50 0000000045f1e6a9 ffff8304a23d8000 ffff8304a480fae0
(XEN) [   12.598843]    ffff8304a4817980 ffff82d040319c11 ffff8304a4817998 ffff8304a4817b70
(XEN) [   12.607709]    6573655220455844 326d657473795374 547465736552203a 646c6f4320657079
(XEN) [   12.616565]    44206c6c6143202c 31203d2068747065 ffff8304000a0d2e 0000000045f2d7c0
(XEN) [   12.625422]    ffff8304a4817a10 0000000045f2b901 ffff8304a4817a10 0000000045f2b3c5
(XEN) [   12.634280]    00000000004a4800 ffff8304a4817a48 0000000000000000 0000000000000007
(XEN) [   12.643137]    ffff8304a4817b50 0000000045f2bd90 0000000000000046 ffff8304a480fb18
(XEN) [   12.651995]    ffff8304a4800f10 ffff8304a480fae0 ffff8304a4817a58 ffff82d04023110e
(XEN) [   12.660853]    ffff82d040340ff7 ffff82d0405ba4a0 ffff8304a4817a70 ffff82d0403412e6
(XEN) [   12.669710]    0000000000000000 0000000000000000 0000000000000000 0000000000000000
(XEN) [   12.678567]    0000000000000000 0000000000000000 0000000000000000 0000000000000000
(XEN) [   12.687422]    0000000000000000 0000000000000000 0000000000000000 0000000000000000
(XEN) [   12.696282]    0000000000000000 0000000000000000 0000000000000000 0000000000000000
(XEN) [   12.705140]    0000000000000000 0000000000000000 0000000000000000 0000000000000000
(XEN) [   12.713999]    0000000000000000 0000000000000000 0000000000000000 0000000000000000
(XEN) [   12.722854]    0000000000000000 0000000000000007 00000004a23ff000 0000000000000000
(XEN) [   12.731709]    ffff8304a4817c70 0000000045f1e999 0000000000000001 ffff83040311100a
(XEN) [   12.740567]    0000000045f20a84 0000000000000001 0000000000000000 ffff82d040231100
(XEN) [   12.749426] Xen call trace:
(XEN) [   12.753044]    [<0000000045f1e11f>] R 0000000045f1e11f
(XEN) [   12.759236]    [<0000000000000000>] S 0000000000000000
(XEN) [   12.765429]    [<0000000045f1e6a9>] F 0000000045f1e6a9
(XEN) [   12.771620]    [<0000000045f1e999>] F 0000000045f1e999
(XEN) [   12.777809]    [<ffff82d040291517>] F efi_reset_system+0x4c/0x6c
(XEN) [   12.784952]    [<ffff82d04033cf6d>] F machine_restart+0x1dd/0x260
(XEN) [   12.792190]    [<ffff82d04033cfff>] F apic_wait_icr_idle+0/0x2f
(XEN) [   12.799241]    [<ffff82d040230c64>] F smp_call_function_interrupt+0xca/0xce
(XEN) [   12.807427]    [<ffff82d04033d8bd>] F call_function_interrupt+0x26/0x38
(XEN) [   12.815239]    [<ffff82d040327369>] F do_IRQ+0x15d/0x713
(XEN) [   12.821623]    [<ffff82d040201902>] F common_interrupt+0x132/0x140
(XEN) [   12.828958]    [<ffff82d040230084>] F rcu_pending+0x5b/0x63
(XEN) [   12.835621]    [<ffff82d0402306c0>] F do_softirq+0x13/0x15
(XEN) [   12.842194]    [<ffff82d040316d5e>] F arch/x86/domain.c#idle_loop+0x92/0xeb
(XEN) [   12.850385] 
(XEN) [   12.852575] Pagetable walk from 00000000465fe004:
(XEN) [   12.858292]  L4[0x000] = 00000004a23fe063 ffffffffffffffff
(XEN) [   12.864864]  L3[0x001] = 00000000436a7063 ffffffffffffffff
(XEN) [   12.871436]  L2[0x032] = 00000000463ff063 ffffffffffffffff
(XEN) [   12.878005]  L1[0x1fe] = 0000000000000000 ffffffffffffffff
(XEN) [   12.884577] 
(XEN) [   12.886770] ****************************************
(XEN) [   12.892766] Panic on CPU 0:
(XEN) [   12.896388] FATAL PAGE FAULT
(XEN) [   12.900104] [error_code=0000]
(XEN) [   12.903914] Faulting linear address: 00000000465fe004
(XEN) [   12.910007] ****************************************
```


</p>
</details>

**Comment by miczyg1** [comment]
@marmarek thanks. Could you please check if MSI v1.1.3 is also affected. The crash happened on an access to RCX pointer `0x00000000465fe000` + offset, which looks like some coreboot memory.

We had a similar issue before on Z690-A v1.1.2/ Z790-P v0.9.0: https://github.com/Dasharo/dasharo-issues/issues/488 and believe I have addressed it already.

**Comment by marmarek** [comment]
Indeed, looks to be fixed in 1.1.3.

**Comment by marmarek** [comment]
Xen 4.19 (used in R4.3) prefers ACPI reboot also on UEFI (xen-project/xen@5231f97f24ba5b1ff5015130d0d96b93fe7d0988)
